### PR TITLE
feat(exp): update constraints

### PR DESCRIPTION
**Pull request by lorenzogentile404**

### CODE REVIEW

File: exp/constraints.lisp [hunk 1]
```diff
@@ -366,18 +369,18 @@
...
+  (callToISZERO 5 0 (trim_acc)))
 
-(defun (padded_base_2_log)
-  (+ (* 8 (nbytes_excluding_leading_byte)) (nbits_of_leading_byte_excluding_leading_bit)))
+(defun (trim_acc_is_zero)
+  (shift preprocessing/WCP_RES 5))
 
 (defun (trivial_trim)
-  (shift preprocessing/WCP_RES 5))
+  (eq! 1 1))
```
Comment (OlivierBBB) [inline exp/constraints.lisp:383]: What's going on here ? We don't need `trivial_trim` anymore, do we ?

Comment (lorenzogentile404) [reply exp/constraints.lisp:383]: ops, it was a placeholder to compile during WIP. I'll remove it!

File: exp/constraints.lisp [hunk 2]
```diff
@@ -388,35 +391,33 @@
   (begin (if-not-zero (min_cutoff_leq_16)
                       (begin (eq! (shift computation/RAW_ACC -1) (raw_lead_hi))
                              (eq! (shift computation/PLT_JMP -1) (min_cutoff)))
-                      (begin (if-zero (raw_hi_part_is_zero)
-                                      (begin (eq! (shift computation/RAW_ACC -1) (raw_lead_hi))
-                                             (eq! (shift computation/PLT_JMP -1) 16))
-                                      (begin (eq! (shift computation/RAW_ACC -1) (raw_lead_lo))
-                                             (eq! (shift computation/PLT_JMP -1) (- (min_cutoff) 16))))))))
+                      (begin (if-not-zero (raw_lead_hi_is_zero)
```
Comment (OlivierBBB) [inline exp/constraints.lisp:394]: Why are there 2 `(begin ...` ?

Comment (OlivierBBB) [reply exp/constraints.lisp:394]: We can `(force-bin (raw_lead_hi_is_zero))` to make it more efficient.

File: exp/constraints.lisp [hunk 3]
```diff
@@ -388,35 +391,33 @@
...
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;                         ;;
 ;;    5.5 Justify          ;;
 ;;        hub prediction   ;;
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 (defconstraint justify-hub-prediction-modexp-log (:perspective macro :guard IS_MODEXP_LOG)
-  (if-not-zero (trivial_trim)
-               (if-not-zero (raw_hi_part_is_zero)
+  (if-not-zero (min_cutoff_leq_16)
```
Comment (OlivierBBB) [inline exp/constraints.lisp:407]: We can `(force-bin` it

File: exp/constraints.lisp [hunk 4]
```diff
@@ -388,35 +391,33 @@
...
+               (if-not-zero (trim_acc_is_zero)
```
Comment (OlivierBBB) [inline exp/constraints.lisp:408]: We can `(force-bin` it

File: exp/constraints.lisp [hunk 5]
```diff
@@ -388,35 +391,33 @@
...
-                                         (vanishes! (lead_log))
-                                         (eq! (lead_log)
-                                              (* 8 (- (ebs_cutoff) 16)))))
-               (if-not-zero (ebs_cutoff_leq_16)
                             (eq! (lead_log)
                                  (- (padded_base_2_log)
-                                    (* 8 (- 16 (ebs_cutoff)))))
-                            (if-not-zero (raw_hi_part_is_zero)
+                                    (* 8 (- 16 (ebs_cutoff))))))
+               (if-not-zero (raw_lead_hi_is_zero)
```
Comment (OlivierBBB) [inline exp/constraints.lisp:413]: We can `(force-bin` it

File: exp/constraints.lisp [hunk 6]
```diff
@@ -388,35 +391,33 @@
...
+                            (if-not-zero (trim_acc_is_zero)
```
Comment (OlivierBBB) [inline exp/constraints.lisp:414]: We can `(force-bin` it

File: exp/constraints.lisp (hunk 2 above)
Comment (lorenzogentile404) [reply exp/constraints.lisp:394]: There are two `begin` because one is the `if` case and the other is the `else` case

**Comment by OlivierBBB** [comment]
You are using
```rust
(if-not-zero X A B)
```
Where `X` is some shorthand of the form `(some-binary-shorthand)`. They are binary because they are the output of e.g. the WCP module. If instead you use
```rust
(if-not-zero (force-bin X) A B)
```
The resulting constraints are lower degree

**Comment by lorenzogentile404** [comment]
This optimization is already done behind the scene: 

Screenshot 2024-08-12 at 20 22 14

**Pull request merged**